File: server/src/realtime/callRelay.ts
// 通话信令 + 断连 grace 的网关内端实现。逻辑自 utils/socket.ts 的 socket.io 处理器平移,
// 入口从「socket 事件」换成「网关上行透传帧 / 断连通知」,身份从握手 socket 换成网关注入的
// X-User-Id / X-Device-Id 头。状态机(callSession)与下行语义(忙线裁决/属主路由/grace 重连)不变。
//
// 下行投递统一 publish 到 gw:downlink,由网关按 userId / targetDeviceId / exceptDeviceId 代投。

import { buildDownlink } from '../contracts/ws.js';
import { redis } from '../database/redis.js';
import {
  clearSession,
  getSession,
  getUserCall,
  GRACE_MS,
  markAccepted,
  markRejoined,
  markReconnecting,
  tryCreateSession,
} from '../services/callSession.js';
import { decActiveCalls, incActiveCalls, incCallEvent } from '../metrics/metrics.js';

// 把一条下行帧 publish 给指定用户(网关订阅 gw:downlink 后按路由规则代投)。
const publishDownlink = (
  userId: number,
  type: string,
  data: unknown,
  opts?: { targetDeviceId?: string; exceptDeviceId?: string },
): Promise<number> =>
  redis.publish('gw:downlink', JSON.stringify(buildDownlink(userId, type, data, opts)));

type RelayResult = { status: number; body?: unknown };

// 处理一条通话信令上行帧(eventType 由信封 type 给出,event 为信封 data)。
// 返回 HTTP 状态码与响应体:
//  - 204 无回投(网关不回投空帧给发送方);
//  - 200 + 响应帧(仅回投操作端,如 call:busy / 会话已结束的 call:end)。
export async function handleCallEvent(
  eventType: string,
  userId: number,
  deviceId: string,
  event: Record<string, unknown>,
): Promise<RelayResult> {
  const callId = String(event.callId ?? '');

  switch (eventType) {
    case 'call:start': {
      incCallEvent('start');
      const calleeId = Number((event.to as { id?: unknown } | undefined)?.id);
      if (!callId || !Number.isFinite(calleeId) || calleeId <= 0) {
        return { status: 400, body: { type: 'call:error', message: '缺少 callId 或合法被叫' } };
      }
      const ok = await tryCreateSession({
        callId,
        callerId: userId,
        calleeId,
        callType: String(event.callType ?? 'voice'),
        callerDevice: deviceId,
      });
      if (!ok) {
        incCallEvent('busy');
        // 仅回主叫本设备,不打扰被叫。
        return { status: 200, body: { type: 'call:busy', callId } };
      }
      incActiveCalls();
      await publishDownlink(calleeId, 'call:start', event);
      return { status: 204 };
    }

    case 'call:accept': {
      incCallEvent('accept');
      await markAccepted(callId, deviceId);
      // answer 回投给 offer 方;并通知同一被叫用户的其它设备「已在别处接听」,停止振铃。
      const toId = Number(event.to);
      if (Number.isFinite(toId) && toId > 0) {
        await publishDownlink(toId, 'call:accept', event);
      }
      await publishDownlink(userId, 'call:handled', { callId, status: 'accepted' }, {
        exceptDeviceId: deviceId,
      });
      return { status: 204 };
    }

    case 'call:reject': {
      incCallEvent('reject');
      const s = await clearSession(callId);
      // 只有会话确实存在才对称 -1(与 call:start 的 +1 对称)。
      if (s) decActiveCalls();
      const callerId = s ? s.callerId : parseInt(callId.split('_')[1] ?? '', 10);
      if (Number.isFinite(callerId) && callerId > 0) {
        await publishDownlink(callerId, 'call:reject', event);
      }
      await publishDownlink(userId, 'call:handled', { callId, status: 'rejected' }, {
        exceptDeviceId: deviceId,
      });
      return { status: 204 };
    }

    case 'call:end': {
      incCallEvent('end');
      const s = await clearSession(callId);
      // 同上:只有会话确实存在才 -1,与 call:start 的 +1 对称。
      if (s) decActiveCalls();
      const [, u1, u2] = callId.split('_');
      const id1 = parseInt(u1 ?? '', 10);
      const id2 = parseInt(u2 ?? '', 10);
      if (Number.isFinite(id1) && id1 > 0) await publishDownlink(id1, 'call:end', event);
      if (Number.isFinite(id2) && id2 > 0 && id2 !== id1) await publishDownlink(id2, 'call:end', event);
      return { status: 204 };
    }

    case 'call:rejoin': {
      incCallEvent('rejoin');
      const s = await getSession(callId);
      if (!s) {
        // 会话已不存在:让重连方干净收场。
        return { status: 200, body: { type: 'call:end', callId } };
      }
      const side = userId === s.callerId ? 'caller' : 'callee';
      const updated = await markRejoined(callId, side, deviceId);
      const peerDevice = side === 'caller' ? updated?.calleeDevice : updated?.callerDevice;
      const peerId = side === 'caller' ? s.calleeId : s.callerId;
      if (peerDevice) {
        // 精确投给对端属主设备:新 offer 只该设备处理(属主路由)。
        await publishDownlink(peerId, 'call:rejoin', event, { targetDeviceId: peerDevice });
      } else {
        await publishDownlink(peerId, 'call:rejoin', event);
      }
      return { status: 204 };
    }

    case 'call:ice': {
      incCallEvent('ice');
      const [, u1, u2] = callId.split('_');
      const id1 = parseInt(u1 ?? '', 10);
      const id2 = parseInt(u2 ?? '', 10);
      // 转发给双方、排除发送方本设备(socket.to 语义)。
      if (Number.isFinite(id1) && id1 > 0) {
        await publishDownlink(id1, 'call:ice', event, { exceptDeviceId: deviceId });
      }
      if (Number.isFinite(id2) && id2 > 0 && id2 !== id1) {
        await publishDownlink(id2, 'call:ice', event, { exceptDeviceId: deviceId });
      }
      return { status: 204 };
    }

    default:
      return { status: 400, body: { type: 'message.error', message: `不支持的上行类型: ${eventType}` } };
  }
}

// 断连通知:某设备离线。平移 socket.io disconnect 处理器:
// 属主设备掉线 → 会话进入 reconnecting → 通知对端 → GRACE_MS 内无人 rejoin 则结束会话。
export async function handleDisconnect(userId: number, deviceId: string): Promise<void> {
  const callId = await getUserCall(userId);
  if (!callId) return;
  const res = await markReconnecting(callId, deviceId);
  if (!res) return; // 非属主设备(如另开的空闲标签页)掉线,忽略
  const { session, epoch } = res;
  const peerId = userId === session.callerId ? session.calleeId : session.callerId;
  await publishDownlink(peerId, 'call:peer-reconnecting', { callId });
  // grace 到点:重读 Redis,仍处 reconnecting 且 epoch 未变(无人 rejoin)→ 结束。
  // epoch 校验跨副本生效:别处副本的 rejoin 会自增 epoch,使此定时器失效。
  setTimeout(() => {
    void (async () => {
      const cur = await getSession(callId);
      if (cur && cur.status === 'reconnecting' && cur.graceEpoch === epoch) {
        await clearSession(callId);
        await publishDownlink(cur.callerId, 'call:end', { callId });
        await publishDownlink(cur.calleeId, 'call:end', { callId });
      }
    })().catch((err) => console.error('grace 结束处理失败:', err));
  }, GRACE_MS);
}

File: biz/internal/realtime/callRelay.go
package realtime

import (
	"context"
	"encoding/json"
	"log/slog"
	"strconv"
	"strings"
	"time"

	"github.com/our-chat/biz/internal/metrics"
	"github.com/our-chat/biz/internal/service"
)

// callRelay.go:通话信令 + 断连 grace,语义对齐 server/src/realtime/callRelay.ts。
// 下行经 gw:downlink(targetDeviceId/exceptDeviceId 与 buildDownlink 一致)。

// HandleCallEvent 处理一条通话信令上行帧(callRelay.ts:36-148)。
func HandleCallEvent(ctx context.Context, eventType string, uctx UplinkContext, data json.RawMessage) UplinkResult {
	callID := jsonField(data, "callId")

	switch eventType {
	case "call:start": {
		metrics.CallEventsTotal.WithLabelValues("start").Inc()
		calleeID, ok := parseID(jsonField(data, "to"))
		if callID == "" || !ok || calleeID <= 0 {
			return UplinkResult{Status: 400, Body: map[string]any{"type": "call:error", "message": "缺少 callId 或合法被叫"}}
		}
		callType := jsonField(data, "callType")
		if callType == "" {
			callType = "voice"
		}
		created, err := service.TryCreateSession(ctx, &service.CallSession{
			CallID:       callID,
			CallerID:     uctx.UserID,
			CalleeID:     calleeID,
			CallType:     callType,
			Status:       "ringing",
			StartTime:    nil,
			CallerDevice: strPtr(uctx.DeviceID),
			CalleeDevice: nil,
			GraceEpoch:   0,
		})
		if err != nil {
			return UplinkResult{Status: 500, Body: map[string]any{"type": "call:error", "message": "信令处理失败"}}
		}
		if !created {
			metrics.CallEventsTotal.WithLabelValues("busy").Inc()
			// 仅回主叫本设备,不打扰被叫(callRelay.ts:58-62)
			return UplinkResult{Status: 200, Body: map[string]any{"type": "call:busy", "data": map[string]any{"callId": callID}}}
		}
		metrics.ActiveCalls.Inc()
		_ = service.PublishDownlink(ctx, calleeID, "call:start", rawJSON(data), nil)
		return UplinkResult{Status: 204}
	}

	case "call:accept": {
		metrics.CallEventsTotal.WithLabelValues("accept").Inc()
		if _, err := service.MarkAccepted(ctx, callID, uctx.DeviceID); err != nil {
			return UplinkResult{Status: 500, Body: map[string]any{"type": "call:error", "message": "信令处理失败"}}
		}
		toID, _ := parseID(jsonField(data, "to"))
		if toID > 0 {
			_ = service.PublishDownlink(ctx, toID, "call:accept", rawJSON(data), nil)
		}
		_ = service.PublishDownlink(ctx, uctx.UserID, "call:handled",
			map[string]any{"callId": callID, "status": "accepted"},
			&service.DownlinkOpts{ExceptDeviceID: uctx.DeviceID})
		return UplinkResult{Status: 204}
	}

	case "call:reject": {
		metrics.CallEventsTotal.WithLabelValues("reject").Inc()
		s, err := service.ClearSession(ctx, callID)
		if err != nil {
			return UplinkResult{Status: 500, Body: map[string]any{"type": "call:error", "message": "信令处理失败"}}
		}
		if s != nil {
			metrics.ActiveCalls.Dec()
		}
		callerID := int64(0)
		if s != nil {
			callerID = s.CallerID
		} else if parts := strings.Split(callID, "_"); len(parts) > 1 {
			callerID, _ = strconv.ParseInt(parts[1], 10, 64)
		}
		if callerID > 0 {
			_ = service.PublishDownlink(ctx, callerID, "call:reject", rawJSON(data), nil)
		}
		_ = service.PublishDownlink(ctx, uctx.UserID, "call:handled",
			map[string]any{"callId": callID, "status": "rejected"},
			&service.DownlinkOpts{ExceptDeviceID: uctx.DeviceID})
		return UplinkResult{Status: 204}
	}

	case "call:end": {
		metrics.CallEventsTotal.WithLabelValues("end").Inc()
		s, err := service.ClearSession(ctx, callID)
		if err != nil {
			return UplinkResult{Status: 500, Body: map[string]any{"type": "call:error", "message": "信令处理失败"}}
		}
		if s != nil {
			metrics.ActiveCalls.Dec()
		}
		parts := strings.Split(callID, "_")
		if len(parts) > 2 {
			id1, err1 := strconv.ParseInt(parts[1], 10, 64)
			id2, err2 := strconv.ParseInt(parts[2], 10, 64)
			if err1 == nil && id1 > 0 {
				_ = service.PublishDownlink(ctx, id1, "call:end", rawJSON(data), nil)
			}
			if err2 == nil && id2 > 0 && id2 != id1 {
				_ = service.PublishDownlink(ctx, id2, "call:end", rawJSON(data), nil)
			}
		}
		return UplinkResult{Status: 204}
	}

	case "call:rejoin": {
		metrics.CallEventsTotal.WithLabelValues("rejoin").Inc()
		s, err := service.GetCallSession(ctx, callID)
		if err != nil {
			return UplinkResult{Status: 500, Body: map[string]any{"type": "call:error", "message": "信令处理失败"}}
		}
		if s == nil {
			// 会话已不存在:让重连方干净收场(callRelay.ts:113-116)
			return UplinkResult{Status: 200, Body: map[string]any{"type": "call:end", "data": map[string]any{"callId": callID}}}
		}
		side := "callee"
		peerID := s.CallerID
		if uctx.UserID == s.CallerID {
			side = "caller"
			peerID = s.CalleeID
		}
		updated, err := service.MarkRejoined(ctx, callID, side, uctx.DeviceID)
		if err != nil {
			return UplinkResult{Status: 500, Body: map[string]any{"type": "call:error", "message": "信令处理失败"}}
		}
		peerDevice := ""
		if updated != nil {
			if side == "caller" {
				if updated.CalleeDevice != nil {
					peerDevice = *updated.CalleeDevice
				}
			} else if updated.CallerDevice != nil {
				peerDevice = *updated.CallerDevice
			}
		}
		if peerDevice != "" {
			// 精确投给对端属主设备(属主路由,callRelay.ts:121-123)
			_ = service.PublishDownlink(ctx, peerID, "call:rejoin", rawJSON(data),
				&service.DownlinkOpts{TargetDeviceID: peerDevice})
		} else {
			_ = service.PublishDownlink(ctx, peerID, "call:rejoin", rawJSON(data), nil)
		}
		return UplinkResult{Status: 204}
	}

	case "call:ice": {
		metrics.CallEventsTotal.WithLabelValues("ice").Inc()
		parts := strings.Split(callID, "_")
		if len(parts) > 2 {
			id1, err1 := strconv.ParseInt(parts[1], 10, 64)
			id2, err2 := strconv.ParseInt(parts[2], 10, 64)
			if err1 == nil && id1 > 0 {
				_ = service.PublishDownlink(ctx, id1, "call:ice", rawJSON(data),
					&service.DownlinkOpts{ExceptDeviceID: uctx.DeviceID})
			}
			if err2 == nil && id2 > 0 && id2 != id1 {
				_ = service.PublishDownlink(ctx, id2, "call:ice", rawJSON(data),
					&service.DownlinkOpts{ExceptDeviceID: uctx.DeviceID})
			}
		}
		return UplinkResult{Status: 204}
	}

	default:
		return UplinkResult{Status: 400, Body: map[string]any{"type": "message.error", "message": "不支持的上行类型: " + eventType}}
	}
}

// HandleDisconnect 断连通知:属主设备掉线 → reconnecting → 通知对端 → grace 窗内无人 rejoin 则结束。
// (callRelay.ts:152-172;epoch 校验使在途 grace 定时器跨副本失效。)
func HandleDisconnect(ctx context.Context, userID int64, deviceID string, logger *slog.Logger) {
	callID, err := service.GetUserCall(ctx, userID)
	if err != nil || callID == "" {
		return
	}
	s, err := service.MarkReconnecting(ctx, callID, deviceID)
	if err != nil || s == nil {
		return // 非属主设备掉线,忽略(callRelay.ts:156)
	}
	peerID := s.CalleeID
	if userID == s.CalleeID {
		peerID = s.CallerID
	}
	_ = service.PublishDownlink(ctx, peerID, "call:peer-reconnecting", map[string]any{"callId": callID}, nil)

	// grace 到点:重读 Redis,仍 reconnecting 且 epoch 未变(无人 rejoin)→ 结束
	epoch := s.GraceEpoch
	go func() {
		time.Sleep(GraceMS * time.Millisecond)
		cur, err := service.GetCallSession(context.Background(), callID)
		if err != nil {
			logger.Warn("grace 重读会话失败", "callId", callID, "err", err)
			return
		}
		if cur != nil && cur.Status == "reconnecting" && cur.GraceEpoch == epoch {
			cleared, err := service.ClearSession(context.Background(), callID)
			if err != nil {
				logger.Warn("grace 清理会话失败", "callId", callID, "err", err)
				return
			}
			if cleared != nil {
				_ = service.PublishDownlink(context.Background(), cleared.CallerID, "call:end",
					map[string]any{"callId": callID}, nil)
				_ = service.PublishDownlink(context.Background(), cleared.CalleeID, "call:end",
					map[string]any{"callId": callID}, nil)
			}
		}
	}()
}

// GraceMS grace 宽限窗(与 service.GraceMS 一致)。
const GraceMS = service.GraceMS

// ---- helpers ----

func jsonField(data json.RawMessage, key string) string {
	var m map[string]json.RawMessage
	if err := json.Unmarshal(data, &m); err != nil {
		return ""
	}
	raw, ok := m[key]
	if !ok {
		return ""
	}
	// 尝试字符串
	var s string
	if err := json.Unmarshal(raw, &s); err == nil {
		return s
	}
	// 尝试数字(JSON number)
	var f float64
	if err := json.Unmarshal(raw, &f); err == nil {
		return strconv.FormatInt(int64(f), 10)
	}
	return string(raw)
}

// parseID 解析 to 字段:可能直接是数字,也可能是对象 {id}。
func parseID(field string) (int64, bool) {
	if field == "" {
		return 0, false
	}
	if strings.HasPrefix(field, "{") {
		var obj struct {
			ID any `json:"id"`
		}
		if err := json.Unmarshal([]byte(field), &obj); err != nil {
			return 0, false
		}
		switch t := obj.ID.(type) {
		case float64:
			return int64(t), true
		case string:
			n, err := strconv.ParseInt(t, 10, 64)
			return n, err == nil
		}
		return 0, false
	}
	n, err := strconv.ParseInt(field, 10, 64)
	return n, err == nil
}

func strPtr(v string) *string { return &v }

// rawJSON 把原始 JSON 对象直接作为 payload 传出(信令帧原样转发,callRelay.ts 同语义)。
func rawJSON(data json.RawMessage) map[string]json.RawMessage {
	var m map[string]json.RawMessage
	if err := json.Unmarshal(data, &m); err != nil {
		return nil
	}
	return m
}

// parseNonnegAny 非负整数宽松解析。
func parseNonnegAny(v any) (int64, bool) {
	switch t := v.(type) {
	case float64:
		if t >= 0 && t == float64(int64(t)) {
			return int64(t), true
		}
		return 0, false
	case string:
		n, err := strconv.ParseInt(t, 10, 64)
		return n, err == nil && n >= 0
	default:
		return 0, false
	}
}

func logWarn(msg string, err error)  { slog.Default().Warn(msg, "err", err) }
func logDownlinkWarn(err error)      { slog.Default().Warn("下行 publish 失败", "err", err) }
